2.0-proposal/VISION.md

```diff
@@ -0,0 +1,24 @@
+# What is charms.reactive and where do we want to go?
+
+## Goals
+
+ - **Layers** let you build on the work of other charmers.
+ - **Interface Layers** define the protocol for communication between Charms.
+ - **Flags** are used for communication between layers, and to respond to certain conditions.
+
+Charms.reactive gives the community the tools to solve their problems. Challenges should be solved *with* the framework, not *in* the framework. This means that charms.reactive is as lean as possible.
+
+**charms.reactive is not a config management tool.** It is a tool for coordinating config management.
+
+ - charms.reactive is used to figure out what state you want to be in.
+ - Config management is used to get to that state.
+
+*The combination of charms.reactive + a config management tool like puppet is great. charms.reactive creates the puppet manifest describing what state we want to be in and calls Puppet to get us into that state.*
+
+## Usability
+
+It is important that using charms.reactive is easy. This translates into a number of usability goals.
+
+ - The Juju lifecycle underneath charms.reactive is completely hidden from the user.
```

Comment:
> debug-reactive

I think we should be able to create this as a plugin, like what `juju dhx` does.

> Not understanding it gives way to 'magic' and allows concepts like 'sleeping' and 'waking' of the charm to come in.

I'm not convinced that this is true. Programming is all about creating good abstractions. If we can create a good abstraction on top of the hooks concept, we should do that. Stuff like files and folders are amazing abstractions. Most people use files every day without knowing how they actually work. Iptables is another example with the tables and chains. That's not how the code actually works, but the abstractions are so good that it doesn't matter to users of iptables.

Magic is when stuff happens that you can't explain using the abstractions. My comment further down elaborates a bit more on this, but the tldr is: "It's only safe to hide the hooks if we can explain all the behavior using the reactive concepts we present to the user."